### Chargement des packages et lecture des données

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
from scipy import stats
import statsmodels.api as sm
from sklearn.model_selection import train_test_split

In [2]:
df = (
    pd.read_excel("C:/Users/papou/OneDrive/Bureau/Etude_de_cas/base3.xlsx")
    .rename(columns = {"CalYear":"YEAR","Group1":"POWER","Poldur":"SENIORITY","Adind":"ALLRISK","Group2":"REGION"})
    .rename(columns = lambda x: x.upper())
)

In [3]:
np.random.seed(28); df.sample(20)

YEAR  GENDER TYPE CATEGORY     OCCUPATION   AGE  POWER  BONUS  \
26025  2009    Male    C    Small      Housewife  49.0     12    -50   
643    2009    Male    D   Medium     Unemployed  42.0     14    -30   
23425  2009    Male    E    Large     Unemployed  28.0     12      0   
38315  2010    Male    D   Medium       Employed  37.0     13    -50   
38793  2010    Male    D   Medium        Retired  46.0      9     80   
31107  2009    Male    A   Medium       Employed  29.0     18     90   
60535  2010  Female    D    Large  Self-employed  27.0      5    -50   
6445   2009  Female    D    Small       Employed  19.0     14      0   
13150  2009    Male    B    Large       Employed  50.0      9    -20   
23914  2009    Male    B    Large        Retired  63.0      9    -30   
59684  2010  Female    E   Medium      Housewife  34.0      3    -20   
6864   2009    Male    B    Large  Self-employed  29.0      7     10   
61461  2010  Female    D    Large       Employed  48.0     18    -50   
53613  2010  Female    E   Medium       Employed  23.0     10      0   
11085  2009    Male    A   Medium       Employed  45.0      7    -50   
23394  2009    Male    D   Medium       Employed  25.0     12      0   
26819  2009    Male    A   Medium        Retired  50.0     11    -40   
64195  2010    Male    B    Small       Employed  45.0      2      0   
42878  2010    Male    A   Medium  Self-employed  54.0      9    -50   
54927  2010  Female    A   Medium  Self-employed  26.0     12     40   

       SENIORITY  VALUE  ALLRISK REGION     DENSITY SURV1  
26025          3   9400        0      S   24.137484     1  
643           15  17700        1      P   37.761578     0  
23425          3  19250        0      S   23.105117     0  
38315          3   6810        0      R  272.966995     0  
38793          4   7710        0      L   45.655642     0  
31107          0  21855        1      L   63.521109     1  
60535          0   6820        0      L   69.889265     0  
6445           6   7670        0      S   23.958550     0  
13150          2   1185        0      Q  222.215127     0  
23914          2   7240        1      R  249.811728     0  
59684         11   8880        0      N   94.296038     0  
6864           5  49230        0      Q   98.103729     0  
61461          0  12775        1      M  129.419475     0  
53613          0  15650        0      Q  174.644525     0  
11085          0  14890        0      R  185.851385     0  
23394         11  18655        0      T   20.851226     0  
26819          3   9880        0      L   79.159448     0  
64195          0  11495        1      R  234.675271     0  
42878         12  20715        1      N  139.606205     0  
54927          6   6220        1      N   88.951524     1

### PREMIERE PARTIE

L'objectif de cette étude de cas est de construire une classification du niveau de risque des assurés d'une compagnie d'assurance automobile. Nous devrons établir  une règle de segmentation qui va permettre d'affecter un score à un individu, et ce score va être notre indicateur du niveau de risque. Pour ce faire, nous disposons d'une base de donnée qui enregistre 71219 observations reparties sur deux années, 35737 pour 2009 et 35542 pour 2010. Sur ces observations, sont mesurées des variables permettant de les caractériser:

-   YEAR : Les années de collecte des données, 2009 et 2010
-   GENDER : Le genre des assurés, masculin(Male) et féminin(Female)
-   TYPE : le type du véhicule une lettre de A à F
-   CATEGORY : Précise la catégorie du véhicule, Small | Medium | Large
-   OCCUPATION : Précise la catégorie CSP de l'assuré
-   AGE : Donne l'âge de l'assuré
-   POWER : valeur numérique de 0 à 20 caractérisant la puissance du véhicule
-   BONUS : Coefficient de malus et bonus de l'assuré
-   SENIORITY : Ancienneté du contrat en nombre d'année
-   VALUE : Valeur en unité monétaire du véhicule 
-   ALLRISK : Indicateur de garantie assurance tout-risque(0 = non & 1 = oui)
-   REGION : Donne la localité de résidence de l'assuré lettre de L à U
-   DENSITY : Densité de la zone d'habitation
-   SURV1 : survenu d'au moins un accident(0 = non & 1 = oui)

In [4]:
df.info(); print("-"*127); df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71279 entries, 0 to 71278
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   YEAR        71279 non-null  int64  
 1   GENDER      71279 non-null  object 
 2   TYPE        71279 non-null  object 
 3   CATEGORY    71279 non-null  object 
 4   OCCUPATION  71279 non-null  object 
 5   AGE         71278 non-null  float64
 6   POWER       71279 non-null  int64  
 7   BONUS       71279 non-null  int64  
 8   SENIORITY   71279 non-null  int64  
 9   VALUE       71279 non-null  int64  
 10  ALLRISK     71279 non-null  int64  
 11  REGION      71279 non-null  object 
 12  DENSITY     71279 non-null  float64
 13  SURV1       71279 non-null  object 
dtypes: float64(2), int64(6), object(6)
memory usage: 7.6+ MB
-------------------------------------------------------------------------------------------------------------------------------


YEAR          0
GENDER        0
TYPE          0
CATEGORY      0
OCCUPATION    0
AGE           1
POWER         0
BONUS         0
SENIORITY     0
VALUE         0
ALLRISK       0
REGION        0
DENSITY       0
SURV1         0
dtype: int64

Seul la variable AGE possède une valeur manquante, étant de donnée que nous ne disposant de variables permettant d'identifier de façon unique les individus, il est préférable de supprimer cet individu de la base de données. Cette suppression n'affectera en rien les résultats de notre analyse on est sur une échelle de 1/71279 soit moins de 0,0014% des observations.

In [5]:
df.describe(include = "number")# Sélectionne des variables numérique et calcul quelques statistiques univariées pour chaque variable

YEAR           AGE         POWER         BONUS     SENIORITY  \
count  71279.000000  71278.000000  71279.000000  71279.000000  71279.000000   
mean    2009.498632     41.181879     10.697709     -7.008796      5.472874   
std        0.500002     15.249655      4.687986     48.547774      4.586798   
min     2009.000000     18.000000      1.000000    -50.000000      0.000000   
25%     2009.000000     30.000000      7.000000    -40.000000      1.000000   
50%     2009.000000     40.000000     11.000000    -30.000000      4.000000   
75%     2010.000000     51.000000     14.000000     10.000000      9.000000   
max     2010.000000    688.000000     20.000000    150.000000     15.000000   

              VALUE       ALLRISK       DENSITY  
count  71279.000000  71279.000000  71279.000000  
mean   16418.451297      0.511525    116.944389  
std    10491.294364      0.499871     79.448243  
min     1000.000000      0.000000     14.377142  
25%     8360.000000      0.000000     50.566406  
50%    14550.000000      1.000000     94.364623  
75%    22550.000000      1.000000    171.372936  
max    49995.000000      1.000000    297.385170

La variable ALLRISK étant une variable numérique(à valeurs binaire) et YEAR une variable numérique à deux modalités, nous allons les transformées en variables catégorielles.

In [6]:
df["AGE"].unique()# Affiche les valeurs distinctes de la variable AGE

array([ 25.,  20.,  21.,  33.,  26.,  29.,  31.,  35.,  27.,  65.,  24.,
        22.,  18.,  34.,  41.,  60.,  62.,  37.,  36.,  46.,  45.,  47.,
        40.,  48.,  74.,  43.,  57.,  42.,  30.,  69.,  66.,  51.,  52.,
        38.,  19.,  50.,  73.,  72.,  71.,  28.,  54.,  63.,  49.,  59.,
        39.,  61.,  23.,  56.,  32.,  53.,  58.,  67.,  55.,  75.,  44.,
        70.,  64.,  68., 399.,  nan, 311., 655., 544., 455., 344., 433.,
       488., 688.])

Deux groupes d'individus qu'on observe ici, ceux avec moins de 100 ans dont l'âge maximal est de 75 ans et ceux avec plus de 100 ans dont les valeurs sont dues à des erreurs d'écritures, nous allons conserver pour ces derniers, que les deux premiers éléments.

In [7]:
for col in df.select_dtypes(include = "object").columns:
    print(df[col].value_counts())
    print("-"*127)

GENDER
Male      45147
Female    26111
M            14
F             7
Name: count, dtype: int64
-------------------------------------------------------------------------------------------------------------------------------
TYPE
A    19724
B    15803
D    14042
C     9806
E     7980
F     3924
Name: count, dtype: int64
-------------------------------------------------------------------------------------------------------------------------------
CATEGORY
Medium    26104
Large     25040
Small     20135
Name: count, dtype: int64
-------------------------------------------------------------------------------------------------------------------------------
OCCUPATION
Employed         22117
Self-employed    14584
Housewife        14235
Unemployed       10969
Retired           9374
Name: count, dtype: int64
-------------------------------------------------------------------------------------------------------------------------------
REGION
L    16909
Q    15911
R    10731
M     5407
P     38

Pour les variables non numériques, nous allons les transformées en variables catégorielle en affectant le groupe de référence à la modalité ayant le nombre d'occurrence le plus élevées. Pour la variable GENDER, les modalité M et F, seront transformée en Male et Female, et pour la variable SURV1 no & yes en 0 et 1 respectivement.

In [8]:
df = (
    df.dropna()
    .assign(
        YEAR = lambda x: x["YEAR"].astype(str),
        ALLRISK = lambda x: x["ALLRISK"].astype(str),
        AGE = lambda x: x["AGE"].astype(str).str[:2].astype(int),
        SURV1 = lambda x: np.select([x["SURV1"] == "yes", x["SURV1"] == "no"], [1,0], default = x["SURV1"]),
        GENDER = lambda x: np.select([x["GENDER"] == "M", x["GENDER"] == "F"],["Male","Female"], default = x["GENDER"])
    )
    .reset_index(drop = True)
)

In [9]:
for col in df.select_dtypes(include = "object").columns:
    print(df[col].value_counts())
    print("-"*127)

df.describe(include = "number")

YEAR
2009    35736
2010    35542
Name: count, dtype: int64
-------------------------------------------------------------------------------------------------------------------------------
GENDER
Male      45160
Female    26118
Name: count, dtype: int64
-------------------------------------------------------------------------------------------------------------------------------
TYPE
A    19723
B    15803
D    14042
C     9806
E     7980
F     3924
Name: count, dtype: int64
-------------------------------------------------------------------------------------------------------------------------------
CATEGORY
Medium    26104
Large     25039
Small     20135
Name: count, dtype: int64
-------------------------------------------------------------------------------------------------------------------------------
OCCUPATION
Employed         22117
Self-employed    14584
Housewife        14235
Unemployed       10968
Retired           9374
Name: count, dtype: int64
--------------------------------

AGE         POWER         BONUS     SENIORITY         VALUE  \
count  71278.000000  71278.000000  71278.000000  71278.000000  71278.000000   
mean      41.121833     10.697719     -7.008895      5.472937  16418.460254   
std       14.291734      4.688018     48.548107      4.586799  10491.367686   
min       18.000000      1.000000    -50.000000      0.000000   1000.000000   
25%       30.000000      7.000000    -40.000000      1.000000   8360.000000   
50%       40.000000     11.000000    -30.000000      4.000000  14550.000000   
75%       51.000000     14.000000     10.000000      9.000000  22550.000000   
max       75.000000     20.000000    150.000000     15.000000  49995.000000   

            DENSITY  
count  71278.000000  
mean     116.943442  
std       79.448397  
min       14.377142  
25%       50.566406  
50%       94.330330  
75%      171.372936  
max      297.385170

In [10]:
np.random.seed(28); df.sample(n = 60)

YEAR  GENDER TYPE CATEGORY     OCCUPATION  AGE  POWER  BONUS  \
32106  2009    Male    A   Medium        Retired   60      9    -40   
61530  2010    Male    A    Large  Self-employed   49     11    -20   
14944  2009    Male    A   Medium       Employed   37     13    -50   
61560  2010    Male    E    Large      Housewife   59     13    -50   
6867   2009  Female    C    Small       Employed   31     19     10   
37786  2010    Male    F    Large      Housewife   43     12    -30   
8979   2009  Female    A    Large      Housewife   53      4    -30   
54426  2010  Female    A   Medium       Employed   36      5    -30   
54320  2010    Male    B    Small     Unemployed   45     13     90   
55596  2010  Female    D   Medium  Self-employed   42      9    -40   
51002  2010    Male    A    Large        Retired   51     11    -50   
47556  2010    Male    D   Medium     Unemployed   37     17    -30   
15579  2009    Male    A    Small     Unemployed   37     11     80   
2795   2009    Male    C    Small  Self-employed   21      8    -10   
3008   2009  Female    E    Small  Self-employed   34     16    -10   
27334  2009  Female    C    Small     Unemployed   37     11    -30   
27023  2009    Male    D    Small  Self-employed   42     11    -40   
21308  2009    Male    A    Small       Employed   22     11      0   
1598   2009  Female    D   Medium     Unemployed   30      7    -40   
5896   2009    Male    A    Large     Unemployed   51     11    -30   
2025   2009  Female    F    Small       Employed   29      8    -10   
18061  2009    Male    E   Medium  Self-employed   37      9    -20   
3628   2009    Male    D    Small        Retired   60      5    110   
48130  2010    Male    B   Medium  Self-employed   27     11     20   
42491  2010    Male    A    Small       Employed   27     15    150   
6483   2009  Female    D    Large       Employed   54     14    -30   
33934  2009    Male    B   Medium       Employed   35     14    -30   
66933  2010  Female    E    Large        Retired   46      9     20   
24519  2009    Male    C    Small  Self-employed   21     13     40   
38002  2010    Male    B    Large      Housewife   50     15     20   
55564  2010  Female    B   Medium      Housewife   39      9     80   
49804  2010    Male    D    Large      Housewife   23      9    -10   
56996  2010    Male    C    Small       Employed   44     15    -50   
3989   2009  Female    A    Small       Employed   30     12    -50   
11933  2009    Male    A    Small  Self-employed   40     11    140   
28562  2009  Female    A   Medium       Employed   53     11    -50   
17781  2009    Male    C   Medium       Employed   39      9    -30   
58394  2010  Female    A   Medium     Unemployed   36      8     20   
42883  2010    Male    B   Medium      Housewife   46     17    -40   
57331  2010    Male    E    Large        Retired   61     12    -40   
26923  2009  Female    D    Small     Unemployed   28      6    -10   
50145  2010    Male    D   Medium  Self-employed   44     13     50   
55321  2010  Female    E    Small      Housewife   26      8    -20   
52438  2010    Male    A    Small       Employed   26      9    -10   
58638  2010    Male    D    Large       Employed   42      4     40   
30589  2009    Male    B    Large  Self-employed   46     10    -50   
34846  2009    Male    B    Large     Unemployed   58     11    -10   
27012  2009  Female    A    Small  Self-employed   21      2     70   
7081   2009  Female    E    Small       Employed   23     11      0   
9732   2009    Male    A    Large     Unemployed   55      7    -30   
46844  2010  Female    A   Medium  Self-employed   59      9     50   
1429   2009  Female    D   Medium      Housewife   73     10    -40   
16031  2009  Female    D    Small     Unemployed   25      5    -10   
65024  2010    Male    C   Medium     Unemployed   42     17    110   
55122  2010    Male    C    Large  Self-employed   45     14     30   
63899  2010    Male    F   Medi

En analysant la base de données complet, la variable AGE est celle où l'on a pu observer des problèmes en particulier. Principalement car on pouvait la classer selon deux groupes, ceux ayant moins de 100 ans et ceux avec plus de 100 ans.
Pour les individus avec moins de 100 ans, l'âge maximal observer était de 75 ans tandis que pour ceux ayant plus de 100 ans, l'âge était du type (311,344,455...) ans, ce qui ressemble le plus à une erreur d'écriture ainsi donc, on a conserver que les deux premiers éléments de la valeur âge pour les individus ayant plus de 100 ans pour avoir une certaine cohérence.

Ensuite, sur la variable SURV1, on observait les modalités "yes" et "no" plutôt que "1" et "0" portant le nombre de modalités à 4 au lieu de 2. On a donc transformé ces valeurs gênante à leur valeur numérique correspondante.

Un pareil vas avec la variable GENDER où sur certains individus, "Male" et "Female" était réduit à "M" et "F", on les donc transformé aux valeurs correspondante pour ramener le nombre de modalités à 2.

A l'issue de ces deux manipulations, on est passé d'une base de données de 71279 observations à 71278 observations soit une suppression de 1 seul individu représentant moins de 0.0014% de l'effectif Total.

In [11]:
X,y = df.drop(columns = ["YEAR","SURV1"]),df["SURV1"]

In [12]:
X_train,X_test,y_train,y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

In [13]:
np.random.seed(28); print(y_train.sample(20)); print("-"*127); X_train.sample(20)

18337    0
22787    0
33170    1
52832    0
39151    0
19526    0
32167    0
32358    0
61997    0
43420    0
26881    1
66791    0
45019    0
3353     1
59650    1
63768    0
25377    0
337      0
53015    0
54564    1
Name: SURV1, dtype: object
-------------------------------------------------------------------------------------------------------------------------------


GENDER TYPE CATEGORY     OCCUPATION  AGE  POWER  BONUS  SENIORITY  \
35953  Female    B    Large        Retired   63     18    -40          0   
24725  Female    C    Large       Employed   24     12     40         11   
57612  Female    A   Medium       Employed   22     11     30          2   
52917  Female    A   Medium     Unemployed   22     12     60         13   
45615    Male    A    Large        Retired   56      3    -40          3   
10283    Male    E    Small      Housewife   39      3     60          9   
33617  Female    D    Large  Self-employed   46     13     30          2   
21352  Female    B    Small       Employed   26     10    -10          0   
14464    Male    B    Small       Employed   31     10    -30          3   
7432     Male    B   Medium     Unemployed   40     11    -40         12   
2749     Male    B    Large        Retired   55     12    -40          4   
54526    Male    C    Small       Employed   30      5     60          1   
57460    Male    D    Small  Self-employed   19      5      0          8   
11063    Male    B   Medium  Self-employed   39      3    -50          3   
9855     Male    C    Small      Housewife   26      8     30          0   
12228  Female    D   Medium     Unemployed   36     10    -50          1   
52141    Male    B    Large        Retired   53      3    -30          8   
24672    Male    A   Medium  Self-employed   31     15    -30          5   
23253    Male    E   Medium  Self-employed   37      8     10          6   
5143     Male    B   Medium        Retired   64      6    -40          2   

       VALUE ALLRISK REGION     DENSITY  
35953  12995       1      N  139.606205  
24725  15450       1      T   25.714011  
57612  11855       1      Q  147.969056  
52917   8905       0      L   91.550641  
45615   5745       1      M  139.283001  
10283   9705       0      Q  206.230276  
33617  44765       1      P   37.837282  
21352   5895       0      R  275.282263  
14464   8805       1      P   33.260934  
7432   21430       1      R  238.240414  
2749    4815       0      N  133.288643  
54526   2135       1      L   67.424787  
57460   8600       0      P   28.951391  
11063  12645       1      U   59.403785  
9855   11490       0      L   84.229038  
12228  21100       1      Q  212.424705  
52141  21530       1      U  121.587022  
24672   7195       1      R  222.711392  
23253   5465       0      L   76.899870  
5143    2990       1      L   76.505375

In [14]:
np.random.seed(28); print(y_test.sample(20)); print("-"*127); X_test.sample(20)

2758     0
5770     0
39289    0
17764    0
71178    0
7875     0
68636    0
26018    1
43808    0
38810    0
56687    0
11583    0
36552    0
60320    0
430      0
16461    0
17662    0
12048    0
70307    1
4699     0
Name: SURV1, dtype: object
-------------------------------------------------------------------------------------------------------------------------------


GENDER TYPE CATEGORY     OCCUPATION  AGE  POWER  BONUS  SENIORITY  \
58082  Female    A    Small  Self-employed   49     10      0          4   
25116    Male    F    Large       Employed   53     20     80          1   
1422     Male    B    Large     Unemployed   20     13     40          8   
50915    Male    A    Large  Self-employed   19     13      0          1   
17485    Male    B    Small      Housewife   35     15    110          0   
11077    Male    A    Large  Self-employed   35     13    -50         13   
32426  Female    B    Small       Employed   27      8     80          4   
11451    Male    B    Large       Employed   51     15    -40          0   
12680  Female    A    Small      Housewife   36     16     40          7   
36922    Male    A    Small       Employed   28     19     30          9   
51165    Male    F   Medium     Unemployed   27      8     30          0   
57845  Female    B    Small     Unemployed   18     14      0         12   
63650  Female    A   Medium     Unemployed   18      8      0          0   
49953  Female    A    Small      Housewife   33     12     30         11   
8420   Female    B    Small      Housewife   21     19      0          2   
70743  Female    A    Large     Unemployed   48     16     80          4   
19349  Female    E    Small  Self-employed   48     11    -50          1   
52303  Female    D   Medium       Employed   31     14    100          7   
57578  Female    B    Small      Housewife   55      7    -30          2   
14966    Male    A    Small       Employed   39     12    -50          7   

       VALUE ALLRISK REGION     DENSITY  
58082   9110       1      M  184.478517  
25116  35410       1      N   88.896231  
1422   17815       0      L   69.728196  
50915   1780       0      R  250.841326  
17485  14775       0      S   17.879958  
11077  24785       1      Q  169.788555  
32426  21465       0      Q  208.666718  
11451   9440       0      T   28.964593  
12680   8455       1      T   25.714011  
36922   5990       1      R  222.711392  
51165   9140       1      O   23.460290  
57845   4640       0      P   33.671239  
63650  17515       0      R  223.347557  
49953   4970       1      Q  114.146701  
8420   44605       0      Q  139.205670  
70743   6600       1      L   51.091482  
19349  14480       1      Q  122.449695  
52303   8870       0      Q  148.494514  
57578  21230       1      L   62.721274  
14966  24260       1      Q  185.134539

### DEUXIEME PARTIE

Les données d'apprentissage qu'on appelé train sont composées de 49875 observations et 14 variables soit 70% des données de df contenant 71279 observations et 14 variables. Toutes les manipulations et mise en forme ont préalablement été faites sur df, train est donc prête à l'emploi.
df est une base de données d'une compagnie d'assurance pour les années 2009 et 2010. Les variables de train ont été décrite après le 4e chunk.
test représente les 30% de restant de df, ce sont les données qui vont nous permettre de juger de la qualité de notre modèle.